core/cache/resident.go

```diff
@@ -0,0 +1,168 @@
+}
+
+// residentManager creates and tracks cache residents.
+// It is also the source for resource identifiers in the cache.
+type residentManager struct {
+	residentCount *counter
+	resourceCount *counter
+
+	residents map[uint64]*resident
+}
+
+func newResidentManager() *residentManager {
+	residentC := counter(0)
+	resourceC := counter(0)
+
+	return &residentManager{
+		residentCount: &residentC,
+		resourceCount: &resourceC,
+		residents:     make(map[uint64]*resident),
+	}
+}
+
+// new creates a uniquely identified type-agnostic cache resident,
+// registers it in the internal map, then returns it.
+func (m *residentManager) new() *resident {
+	r := &resident{
+		id:      m.residentCount.next(),
+		manager: m,
+		workers: make(map[uint64]worker.Worker),
+	}
+	m.residents[r.id] = r
+	return r
+}
+
+func (m *residentManager) evict(id uint64) {
+	// TODO (manadart 2019-04-17): TBC when the mark/sweep logic is added.
+}
+
+func (m *residentManager) deregister(id uint64) {
+	delete(m.residents, id)
+}
+
+// resident is the base class for entities managed in the cache.
+type resident struct {
+	// id uniquely identifies this resident among all
+	// that were supplied by the same resident manager.
+	id uint64
+
+	// stale indicates that this cache resident is stale
+	// and is a candidate for removal.
+	stale bool
+
+	// manager is the resident manager that is responsible for this resource.
+	// If the resident is being evicted, it deregisters from the manager.
+	// TODO (manadart 2019-04-16): Maybe?
+	manager *residentManager
+
+	// workers are resources that must be cleaned up when a resident is to be
+	// evicted from the cache.
+	// Obvious examples are watchers created by the resident.
+	// Access to this map should be Goroutine-safe.
+	workers map[uint64]worker.Worker
+	mu      sync.Mutex
+}
+
+// registerWorker is used to indicate that the input worker needs to be stopped
+// when this resident is evicted from the cache.
+// The deregistration method is returned.
+// TODO (manadart 2019-04-16): Handle case where registration is called
+// on a stale resident.
+func (r *resident) registerWorker(w worker.Worker) func() {
+	id := r.manager.resourceCount.next()
+	r.mu.Lock()
+	r.workers[id] = w
+	r.mu.Unlock()
+	return func() { r.deregisterWorker(id) }
+}
+
+// evict cleans up
+func (r *resident) evict() error {
+	if err := r.cleanup(); err != nil {
+		return errors.Trace(err)
+	}
+	r.manager.deregister(r.id)
+	return nil
+}
+
+// cleanup performs all resource maintenance associated with a resident
+// being evicted from the cache.
+// Note that this method does not deregister the resident from the manager.
+func (r *resident) cleanup() error {
+	return errors.Trace(r.cleanupWorkers())
+}
+
+// cleanupWorkers calls "Stop" on all registered workers
+// and removes them from the internal map.
+func (r *resident) cleanupWorkers() error {
+	for id := range r.workers {
+		if err := r.cleanupWorker(id); err != nil {
```

Comment:
Prefer to collate errors, but keep trying to clean up others.
At the outer loop, log the error, but don't stop.